File: C++/code_0503.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <stdio.h>
using namespace std;

const long long INF = 1LL << 60; //2^60 を表す．十分大きい値を設定する．long は int よりも大きい桁の整数を扱うときに使う．long long は longよりも大きい桁の整数を使うときに使う．

template <class T>
void chmin(T &a, T b)
{
    if (a > b)
    {
        a = b;
    }
}

int main()
{
    int N;
    cout << "N = ";
    cin >> N;

    vector<long long> h(N);
    for (int i = 0; i < N; i++)
    {
        cin >> h[i];
    }

    //初期化(最小化問題なので INF に初期化)
    vector<long long> dp(N, INF);

    //初期条件
    dp[0] = 0;

    //ループ
    for (int i = 1; i < N; i++)
    {
        //cout << "i = " << i << endl;

        //一旦ここで処理がある
        chmin(dp[i], dp[i - 1] + abs(h[i] - h[i - 1]));

        //cout << "dp[" << i << "] = " << dp[i] << endl;

        // i > 1の時，以下の処理もする．
        // 配列dpが配列外参照を起こさないようにする．
        if (i > 1)
        {
            chmin(dp[i], dp[i - 2] + abs(h[i] - h[i - 2]));
            //cout << "dp[" << i << "] = " << dp[i] << endl;
        }
        //cout << "dp[" << i << "] = " << dp[i] << endl;

        //cout << "-----" << endl;
    }

    //答え
    //cout << dp[N - 1] << endl;

    for (int i = 0; i < N; i++)
    {
        cout << "dp[" << i << "] = " << dp[i] << endl;
    }
}

File: Julia/code_0503.jl
function chmin(a, b)
    println("before chmin : chmin(a , b) =  chmin($a , $b)")
    if a > b
        a = b
    end
    println("after chmin : a = $a , b = $b")
    return a, b
end

INF = 2^60

h = [2,9,4,5,1,6,10]
N = length(h)

dp = Vector{Int64}(undef, N)

for i in 1:N
    dp[i] = INF
end

# 初期条件
dp[1] = 0

for i in 2:N
    println("i = $i")
    println("-- 1つ前からの移動 --")
    println("-- before chmin --")

    value = dp[i]
    cost = abs(h[i] - h[i - 1])
    sum = dp[i - 1] + cost

    println("value = $value , cost = $cost , sum = $sum")

    chmin(value, sum)

    """
    chmin の結果に関係なく value = dp[i] の値を更新．
    chmin 関数内では更新してくれないから．
    """

    dp[i] = chmin(value, sum)[1]

    println("-- after chmin --")
    value = dp[i]
    println("value = $value")

    if i > 2
        println("--2つ前からの移動 --")
        value = dp[i]
        cost = abs(h[i] - h[i - 2])
        sum = dp[i - 2] + cost

        println("value = $value , cost = $cost , sum = $sum")

        chmin(value, sum)

        """
        i-1 番目からの総和が小さければここの処理で値は変わらない．
        i-2 番目からの総和の方が小さければここで書き換える．
        """

        dp[i] = chmin(value, sum)[1]

        println("-- after chmin --")
        value = dp[i]
        println("value = $value")
    end

    println("$i 番目に到達するためのコストの総和の最小値 = $value")
    println("------")
end


# 答えの確認
for i in 1:N
    result = dp[i]
    println("dp[$i] = $result")
end
